pkg/virt-controller/network/vmcontroller_test.go

```diff
@@ -0,0 +1,328 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+
+package network_test
+
+import (
+	"context"
+	"errors"
+
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	k8sv1 "k8s.io/api/core/v1"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	"kubevirt.io/client-go/generated/kubevirt/clientset/versioned/fake"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+	"kubevirt.io/kubevirt/pkg/virt-controller/network"
+)
+
+var _ = Describe("VM Network Controller", func() {
+	emptyVM := &v1.VirtualMachine{Spec: v1.VirtualMachineSpec{Template: &v1.VirtualMachineInstanceTemplateSpec{}}}
+
+	It("sync does nothing when the hotplug FG is unset", func() {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{}, stubPodGetter{})
+		originalVM := emptyVM.DeepCopy()
+		Expect(c.Sync(emptyVM, libvmi.New())).To(Equal(originalVM))
+	})
+
+	DescribeTable("sync does nothing when", func(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstance, podGetter stubPodGetter) {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{netHotplugEnabled: true}, podGetter)
+		originalVM := emptyVM.DeepCopy()
+		Expect(c.Sync(vm, vmi)).To(Equal(originalVM))
+	},
+		Entry("the VM is not running (there is no VMI)", emptyVM, nil, stubPodGetter{}),
+		Entry(
+			"the VMI is marked for deletion",
+			emptyVM,
+			&v1.VirtualMachineInstance{ObjectMeta: k8smetav1.ObjectMeta{DeletionTimestamp: &k8smetav1.Time{}}},
+			stubPodGetter{},
+		),
+		Entry("the VM & VMI have no interfaces and no pod is found", emptyVM, libvmi.New(), stubPodGetter{}),
+	)
+
+	It("sync fails when pod fetching returns an error", func() {
+		c := network.NewVMNetController(
+			fake.NewSimpleClientset(),
+			stubClusterConfig{netHotplugEnabled: true},
+			stubPodGetter{err: errors.New("test")},
+		)
+		originalVM := emptyVM.DeepCopy()
+		updatedVM, err := c.Sync(emptyVM, libvmi.New())
+		Expect(err).To(MatchError(isSyncErrorType, "syncError"))
+		Expect(updatedVM).To(Equal(originalVM))
```

Comment:
Isn't a value returned from a function that also returned an error is undefined?
Same bellow.

Reply:
This is how the current code behaves. It returns the current VM in case of failure.
I do not want to change this behavior as part of this PR.

Also, as a caller, I guess you can ignore the value. But in tests, we are suppose to express what actually happens. The fact it can contain information, I think it is important to express it.